### PR TITLE
feat: set Flavor on enroll

### PR DESCRIPTION
**Pull request by skrobul**
This PR enhances the "enroll-server" workflow with a capability of setting the baremetal node's flavour. We have previously attempted to do this through Ironic inspection hooks, but found that running Ironic's inspection has some undesired side effects (i.e. modifying the baremetal node ports for no reason). As a result, we cannot match the node's hardware specs to a given flavor during the inspection.

What we do instead is retrieve all the required inputs through Redfish interface as part of the 'enroll' workflow, then feed this information to the flavor matching code. The resulting flavor is set immediately when the Ironic node is created or updated.

In order to facilitate this, some additional changes were required:
- the flavor matching code has been extracted into separate, independent module in case we want to later move this functionality back to Ironic. For now this is not possible until inspection issues are addressed upstream.
- [flavor-sync](https://github.com/RSS-Engineering/undercloud-deploy/pull/222) has been moved to a separate Deployment backed by a PVC that is shared with instances of the `enroll-server` workflow.
- when the flavors are matched, we now strip first chunk from the name (i.e. 'prod')

Closes https://rackspace.atlassian.net/browse/PUC-591

### CODE REVIEW

**Comment by skrobul** [comment]
side note: I really don't like how the `pre-commit`/`pyright` setup is done when it comes to installing a "local" package, but this was the only way I could get it to work. Open to any suggestions how to do this better.
The [pyright-python](https://github.com/RobertCraigie/pyright-python?tab=readme-ov-file#pre-commit) page suggests that we should either:
-  [tell pre-commit](https://pre-commit.com/#config-additional_dependencies) to install those dependencies, which is great for published modules, but it [intentionally does not work](https://github.com/pre-commit/pre-commit/issues/1752#issuecomment-754252663) for modules/packages included in the source code.
 
- explicitly tell pyright which virtual environment to use by updating your [pyright configuration file](https://github.com/microsoft/pyright/blob/main/docs/configuration.md) - this may work, but I was hesitant to manipulate this because each user will have different configuration settings with very different paths to their virtualenvs. We may be able to inject ephemeral `pyrightconfig.json` into a workspace, but that would also require us to create a virtualenv earlier and install our packages there. Feels like a hack, hence looking for better options.

File: python/understack-flavor-matcher/tests/test_flavor_spec.py [hunk 1]
```diff
@@ -147,41 +172,73 @@ def flavors():
...
 
 def test_cpu_model_not_matching(flavors):
-    machine = Machine(memory_mb=102400, cpu="Non-Existent CPU Model", disk_gb=500)
+    machine = Machine(
+        memory_mb=102400,
+        cpu="Non-Existent CPU Model",
+        disk_gb=500,
+        model="Dell XPS1319",
+    )
     assert all(flavor.score_machine(machine) for flavor in flavors) == 0
```
Comment (stevekeay) [inline python/understack-flavor-matcher/tests/test_flavor_spec.py:223]: Are these asserts doing what it looks like they should be doing?   all() answers a boolean, which `==` compares as a truthy/falsey, so I think this is asserting that at least ONE of the scores is zero, which is all mighty confusing.

**Review by stevekeay — APPROVED** [review]

File: python/understack-workflows/understack_workflows/bmc_disk.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+from dataclasses import dataclass
+
+from understack_workflows.bmc import Bmc
+from understack_workflows.bmc import RedfishError
+from understack_workflows.helpers import setup_logger
+
+logger = setup_logger(__name__)
+
+
+REDFISH_DISKS_PATH = "/redfish/v1/Systems/System.Embedded.1/Storage/RAID.SL.1-1"
```
[the remark this answers is not in the document]
Comment (skrobul) [reply python/understack-workflows/understack_workflows/bmc_disk.py:11]: Yeah, that can be different even when the disks are connected different way (some of the Dell's examples list `RAID.Mezzanine.1-1` for example. As we discussed in the other PR, I went with hardcoded paths that we can differentiate on later.

File: python/understack-workflows/understack_workflows/bmc_disk.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+
+def physical_disks(bmc: Bmc) -> list[Disk]:
+    """Retrieve list of physical physical_disks."""
+    try:
+        disks = bmc.redfish_request(REDFISH_DISKS_PATH)["Drives"]
+        disk_list = [Disk.from_path(bmc, path=disk["@odata.id"]) for disk in disks]
+        logger.debug("Retrieved %d disks.", len(disk_list))
+        return disk_list
+    except RedfishError as err:
+        logger.error("Failed retrieving disk info from %s: %s", REDFISH_DISKS_PATH, err)
```
[the remark this answers is not in the document]
Comment (skrobul) [reply python/understack-workflows/understack_workflows/bmc_disk.py:51]: Good catch, thank you!

File: python/understack-flavor-matcher/tests/test_flavor_spec.py (hunk 1 above)
Comment (skrobul) [reply python/understack-flavor-matcher/tests/test_flavor_spec.py]: You are absolutely right - they were indeed matching incorrectly - it was mean to be 

```python
assert all(flavor.score_machine(machine) == 0 for flavor in flavors)
```

addressed in [93a3caa](https://github.com/rackerlabs/understack/pull/481/commits/93a3caa9fb17538405b0285f02b1aa5915801274)

**Comment by skrobul** [comment]
Merge when the https://github.com/RSS-Engineering/undercloud-deploy/pull/222/ is reviewed and approved

**Pull request merged**